modules/loader/source/relocations.cpp

```diff
@@ -645,88 +650,261 @@ bool resolveAArch64(const loader::Relocation &r, loader::ElfMap &map,
   }
   return true;
 }
+
+// Sets the immediate field of an I-type instruction [31:20]
+void writeITypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 20, 12);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+// Sets the immediate field of an U-type instruction [31:12]
+void writeUTypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 12, 20);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+uint32_t getLo12(uint64_t addr) {
+  return static_cast<uint32_t>(getBitRange(addr, 0, 12));
+}
+
+std::optional<uint32_t> getHi20(int64_t addr) {
```

Comment:
What does this do, why do we sign extend an addr - is it a relative offset rather than an addr?

Reply:
No, this is generic enough for either absolute (`HI20`) or relative (`PCREL_HI20`), so `addr` is an address that we're relocating. However, a `HI20` or `PCREL_HI20` relocation is always a 32-bit address, so we sign extend the 64-bit address we have allocated to assert loudly we're not incorrectly resolving the relocation to a symbol that's too big (absolute) or too far away (relative) for the instruction using it.

Reply:
if it's absolute can't the top bit be set as an unsigned value?

Reply:
Hmm yeah that sounds like I've got a bug after I started refactoring things together, thanks.

Reply:
On a second look I don't think this is a bug because all of the `HI20` relocations use the `LUI` or `AUIPC` instructions which sign-extend the immediate up to 32 or 64 bits depending on the RISC-V architecture. So a 32-bit absolute address with the top bit set on `riscv64` would necessarily be sign-extended to the incorrect address. I think we are correct to throw an error on that (it appears that `lld` does the same thing, fwiw)